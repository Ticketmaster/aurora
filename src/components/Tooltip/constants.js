export const SPACE_FROM_MOUSE = 20;

export const directions = {
  top: "top",
  bottom: "bottom",
  left: "left",
  right: "right"
};

export const variants = {
  dark: "dark",
  light: "light"
};
