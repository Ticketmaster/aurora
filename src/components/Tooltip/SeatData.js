import React from "react";
import PropTypes from "prop-types";

import {
  SeatDataStyled,
  SeatDataColumn,
  ColumnHeading,
  ColumnText
} from "./Tooltip.styles";

import { SIZES_SL, LARGE, VARIANTS, LIGHT } from "../constants";

const SeatData = ({ size, isLast, section, row, seat, variant }) => (
  <SeatDataStyled size={size} isLast={isLast}>
    <SeatDataColumn>
      <ColumnHeading size={size}>Sec</ColumnHeading>
      <ColumnText variant={variant} size={size}>
        {section}
      </ColumnText>
    </SeatDataColumn>
    <SeatDataColumn>
      <ColumnHeading size={size}>Row</ColumnHeading>
      <ColumnText variant={variant} size={size}>
        {row}
      </ColumnText>
    </SeatDataColumn>
    <SeatDataColumn>
      <ColumnHeading size={size}>Seat</ColumnHeading>
      <ColumnText variant={variant} size={size}>
        {seat}
      </ColumnText>
    </SeatDataColumn>
  </SeatDataStyled>
);

SeatData.propTypes = {
  section: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  row: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  seat: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  size: PropTypes.oneOf(SIZES_SL),
  variant: PropTypes.oneOf(VARIANTS),
  isLast: PropTypes.bool
};

SeatData.defaultProps = {
  size: LARGE,
  variant: LIGHT,
  isLast: false
};

SeatData.displayName = "SeatData";

export default SeatData;
