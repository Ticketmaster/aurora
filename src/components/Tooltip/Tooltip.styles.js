import styled from "styled-components";
import { themes, constants, spacing, typography } from "../../theme";
import { popContainersBoxShadow } from "../../theme/constants";
import { directions, variants, seatTooltipSizes } from "./constants";

export const StyledTooltip = styled.div`
  background-color: ${({ variant }) =>
    variant === variants.dark
      ? themes.global.darkFill
      : themes.global.white.base};
  border: ${({ variant }) =>
    variant === variants.dark
      ? `1px solid ${themes.global.darkFill}`
      : `1px solid ${themes.global.gray02}`};
  border-radius: ${constants.borderRadius.large};
  box-shadow: ${popContainersBoxShadow};
  position: absolute;
  max-width: 260px;
  color: ${({ variant }) =>
    variant === variants.dark
      ? themes.global.white.base
      : themes.global.gray01};
  padding: ${spacing.cozy};
  display: ${({ isVisible }) => (isVisible ? "block" : "none")};
  font-size: ${typography.size.uno};
  transition: opacity 0.1s ${constants.easing.easeInQuad},
    transform 0.1s ${constants.easing.easeInQuad};

  :before {
    content: "";
    position: absolute;
    transition: opacity 0.1s ${constants.easing.easeInQuad},
      scale 0.1s ${constants.easing.easeInQuad};
    display: ${({ isVisible }) => (isVisible ? "inline-block" : "none")};
    border-right: ${({ variant }) =>
      variant === variants.light ? `1px solid ${themes.global.gray02}` : ""};
    border-bottom: ${({ variant }) =>
      variant === variants.light ? `1px solid ${themes.global.gray02}` : ""};
    border-top-left-radius: 100%;
    width: 12px;
    height: 12px;
    transform: translateY(-50%) rotate(-135deg);
    background-color: ${({ variant }) =>
      variant === variants.dark
        ? themes.global.darkFill
        : themes.global.white.base};
    ${({ direction }) => {
      switch (direction) {
        case directions.top:
          return "left: calc(50% - 6px); bottom: -13px; transform: translateY(-50%) rotate(45deg);";
        case directions.bottom:
          return "left: calc(50% - 6px); top: -1px; transform: translateY(-50%) rotate(-135deg);";
        case directions.left:
          return "top: 10px; right: -7px; transform: translateY(0%) rotate(-45deg);";
        case directions.right:
          return "top: 10px; left: -7px; transform: translateY(0%) rotate(135deg);";
        default:
          return "left: calc(50% - 6px); top: -1px; transform: translateY(-50%) rotate(-135deg);";
      }
    }};
  }

  &.open-enter,
  &.open-enter:before {
    transition: opacity 0.3s ${constants.easing.easeInOutQuad};
    display: block;
    opacity: 0;
  }

  &.open-enter {
    transition: opacity 0.3s ${constants.easing.easeInOutQuad},
      transform 0.3s ${constants.easing.easeInOutQuad};
    transform: scale(0.7);
  }

  &.open-enter-active,
  &.open-enter-active:before {
    transition: opacity 0.3s ${constants.easing.easeInOutQuad};
    display: block;
    opacity: 1;
  }

  &.open-enter-active {
    transition: opacity 0.3s ${constants.easing.easeInOutQuad},
      transform 0.3s ${constants.easing.easeInOutQuad};
    transform: scale(1);
  }

  &.open-enter-done,
  &.open-enter-active:before {
    transition: opacity 0.3s ${constants.easing.easeInOutQuad};
    display: block;
    opacity: 1;
  }

  &.open-enter-done {
    transition: opacity 0.3s ${constants.easing.easeInOutQuad},
      transform 0.3s ${constants.easing.easeInOutQuad};
    transform: scale(1);
  }

  &.open-exit,
  &.open-exit:before {
    display: block;
    opacity: 1;
  }

  &.open-exit {
    transform: scale(1);
  }

  &.open-exit-active,
  &.open-exit-active:before {
    display: block;
    opacity: 0;
  }

  &.open-exit-active {
    transform: scale(0.7);
  }
`;

export const SeatData = styled.div`
  display: flex;
  justify-content: space-between;
  padding: ${({ size }) =>
    size === seatTooltipSizes.large ? spacing.moderate : spacing.cozy};
  border-bottom: ${({ hasChildren }) =>
    hasChildren ? `1px solid ${themes.global.gray02}` : "none"};
  line-height: 1;
`;

export const SeatDataColumn = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  flex-grow: 1;
  text-align: center;

  &:first-child {
    text-align: left;
  }

  &:last-child {
    text-align: right;
  }
`;

export const ColumnHeading = styled.div`
  font-size: ${({ size }) =>
    size === seatTooltipSizes.large
      ? typography.size.hecto
      : typography.size.uno};
  text-transform: uppercase;
  margin-bottom: ${({ size }) =>
    size === seatTooltipSizes.large ? spacing.cozy : spacing.cozy};
  color: ${themes.global.gray02};
`;

export const ColumnText = styled.div`
  font-size: ${({ size }) =>
    size === seatTooltipSizes.large
      ? typography.size.kilo
      : typography.size.hecto};
  font-weight: ${typography.weight.semiBold};
  color: ${({ variant }) =>
    variant === "dark" ? themes.global.white.base : themes.global.gray01};
`;

export const AdditionalData = styled.div`
  padding: ${({ size }) =>
    size === seatTooltipSizes.large ? spacing.moderate : spacing.cozy};
  font-size: ${({ size }) =>
    size === seatTooltipSizes.large
      ? typography.size.hecto
      : typography.size.uno};
  color: ${({ variant }) =>
    variant === "dark" ? themes.global.white.base : themes.global.gray01};
`;
