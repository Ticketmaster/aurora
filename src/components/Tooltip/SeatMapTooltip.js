import React, { Component } from "react";
import PropTypes from "prop-types";
import { CSSTransition } from "react-transition-group";
import Portal from "../PopOver/PopOverPortal";
import { StyledTooltip } from "./Tooltip.styles";
import SPACE_FROM_MOUSE from "./constants";
import { TOP, BOTTOM, VARIANTS, LIGHT, ARROW_WIDTH } from "../constants";
import { constants } from "../../theme";

class SeatMapTooltip extends Component {
  constructor(props) {
    super(props);

    this.myRef = React.createRef();

    this.dimensions = {
      width: 0,
      height: 0,
      windowScroll: 0,
      windowWidth: 0,
      windowHeight: 0
    };

    this.pos = {
      x: 0,
      y: 0
    };

    this.state = {
      actualDirection: this.props.preferTop ? TOP : BOTTOM,
      arrowAdjustment: 0
    };
  }

  componentDidMount() {
    this.refresh();
  }

  /*
   * If any specific value is passed via `direction` props, than that value should be used
   * If `auto` value is passed via `direction` prop, then calculated direction value will be used
   */
  getDirection = () => {
    const { actualDirection } = this.state;
    return actualDirection;
  };

  getPositionAndUpdateDirection = ({
    position,
    dimensions,
    spaceFromMouse
  }) => {
    const result = this.getTooltipPosition({
      position,
      dimensions,
      spaceFromMouse
    });

    let direction = this.props.preferTop ? TOP : BOTTOM;

    if (result.y < position.y + dimensions.windowScroll) {
      direction = TOP;
    } else if (result.y > position.y + dimensions.windowScroll) {
      direction = BOTTOM;
    }

    this.setState((prevState, props) => {
      const { directionChanged } = props;
      const { actualDirection } = prevState;

      if (actualDirection !== direction) {
        if (direction && directionChanged) {
          directionChanged(direction);
        }

        return {
          actualDirection: direction || actualDirection
        };
      }

      return null;
    });

    return result;
  };

  getTooltipPosition = ({ position, dimensions }) => {
    const { preferTop, spaceFromMouse } = this.props;

    const {
      width,
      windowScroll,
      height,
      windowWidth,
      windowHeight
    } = dimensions;

    const { x, y } = position;

    const viewportBottom = windowScroll + windowHeight;
    const bottomPosition = windowScroll + y + spaceFromMouse;
    const topPosition = windowScroll + y - (spaceFromMouse + height);

    const topPositionWithFallback =
      topPosition > windowScroll ? topPosition : bottomPosition;

    const containerCenter = width / 2;

    const getXPosition = () => {
      if (x - containerCenter < 0) {
        return 0 - ARROW_WIDTH / 2;
      }

      if (x - containerCenter > 0 && x + containerCenter < windowWidth) {
        return x - containerCenter;
      }

      if (x + containerCenter > windowWidth) {
        return windowWidth - width + ARROW_WIDTH / 2;
      }

      return x;
    };

    const getYPosition = () =>
      !preferTop && bottomPosition + height < viewportBottom
        ? bottomPosition
        : topPositionWithFallback;

    this.setArrowPosition(dimensions, x);

    return {
      x: getXPosition(),
      y: getYPosition()
    };
  };

  getTranslateByDirection = direction => {
    switch (direction) {
      case TOP:
        return "translate(0, 10px)";
      case BOTTOM:
        return "translate(0, -10px)";
      default:
        return "translate(0, 10px)";
    }
  };

  setArrowPosition = (dimensions, xPos) => {
    const rightBound = Math.min(
      dimensions.windowWidth,
      dimensions.width + xPos
    );

    const leftBound = 0;
    const rightEdge = xPos + dimensions.width / 2;
    const leftEdge = xPos - dimensions.width / 2;
    const arrowSpace = ARROW_WIDTH / 2;
    const borderRadius = parseInt(
      constants.borderRadius.large.split("px")[0],
      10
    );

    let centerOffset = 0;

    if (leftEdge < leftBound) {
      centerOffset = (leftEdge - leftBound) * 1 + arrowSpace;
    } else if (rightEdge > rightBound) {
      centerOffset = (rightEdge - rightBound) * 1 - (arrowSpace - borderRadius);
    }

    this.setState({
      arrowAdjustment: centerOffset
    });
  };

  /* 
   * Function that forces a re-render of the tooltip
   */

  refresh = () => {
    if (this.props.isVisible) {
      this.tooltipEnter();
      this.tooltipEntering();
    }
  };

  updateSize = () => {
    const { isVisible } = this.props;

    const dimensions = {};

    if (global.window && isVisible) {
      const { innerWidth, innerHeight } = global.window;

      const scrollTop = Math.max(
        global.window.pageYOffset,
        global.document.documentElement.scrollTop,
        global.document.body.scrollTop
      );

      dimensions.windowScroll = scrollTop;
      dimensions.windowWidth = innerWidth;
      dimensions.windowHeight = innerHeight;
    }

    if (this.myRef.current) {
      this.myRef.current.style.top = "0px";
      this.myRef.current.style.left = "0px";

      const { clientWidth, clientHeight } = this.myRef.current;

      dimensions.width = clientWidth;
      dimensions.height = clientHeight;
    }

    if (Object.keys(dimensions).length) {
      this.dimensions = dimensions;
      return true;
    }

    return false;
  };

  tooltipEnter = () => {
    const { isVisible, position, spaceFromMouse } = this.props;

    if (isVisible) {
      this.updateSize();

      this.pos = this.getPositionAndUpdateDirection({
        position,
        dimensions: this.dimensions,
        spaceFromMouse
      });
    }

    this.myRef.current.style.top = `${this.pos.y}px`;
    this.myRef.current.style.left = `${this.pos.x}px`;

    this.myRef.current.style.transition = `opacity 0.3s ${
      constants.easing.easeOutQuad
    }`;

    this.myRef.current.style.transform = this.getTranslateByDirection(
      this.state.actualDirection
    );
  };

  tooltipEntering = () => {
    this.myRef.current.style.transition = `opacity 0.3s ${
      constants.easing.easeOutQuad
    },
      transform 0.3s ${constants.easing.easeOutQuad}`;
    this.myRef.current.style.transform = "translate(0)";
  };

  tooltipExit = () => {
    this.myRef.current.style.transition = `opacity 0.1s ${
      constants.easing.easeInQuad
    }`;
  };

  render() {
    const { children, isVisible, variant, ...rest } = this.props;
    const { arrowAdjustment } = this.state;
    const direction = this.getDirection();

    return (
      <Portal>
        <CSSTransition
          in={isVisible}
          key="tooltip-animation"
          timeout={300}
          classNames="open"
          onEnter={this.tooltipEnter}
          onEntering={this.tooltipEntering}
          onExit={this.tooltipExit}
          appear={isVisible}
          variant={variant}
        >
          <StyledTooltip
            ref={this.myRef}
            isVisible={isVisible}
            {...rest}
            direction={direction}
            arrowAdjustment={`${arrowAdjustment}px`}
          >
            {children}
          </StyledTooltip>
        </CSSTransition>
      </Portal>
    );
  }
}

SeatMapTooltip.propTypes = {
  children: PropTypes.node.isRequired,
  isVisible: PropTypes.bool,
  position: PropTypes.shape({
    x: PropTypes.number,
    y: PropTypes.number
  }),
  variant: PropTypes.oneOf(VARIANTS),
  spaceFromMouse: PropTypes.number,
  preferTop: PropTypes.bool,
  directionChanged: PropTypes.func
};

SeatMapTooltip.defaultProps = {
  isVisible: false,
  variant: LIGHT,
  position: {
    x: 0,
    y: 0
  },
  spaceFromMouse: SPACE_FROM_MOUSE,
  preferTop: true,
  directionChanged: null
};

SeatMapTooltip.displayName = "SeatMapTooltip";

export default SeatMapTooltip;
