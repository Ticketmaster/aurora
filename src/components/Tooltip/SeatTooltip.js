import React, { Component } from "react";
import PropTypes from "prop-types";
import Tooltip from "./index";
import { seatTooltipSizes } from "./constants";
import {
  SeatData,
  SeatDataColumn,
  ColumnHeading,
  ColumnText,
  AdditionalData
} from "./Tooltip.styles";

class SeatTooltip extends Component {
  static getDimensionsFromEvent = e => Tooltip.getDimensionsFromEvent(e);

  render() {
    const { size, row, seat, section, children, ...rest } = this.props;
    const maxWidth = size === seatTooltipSizes.small ? "180px" : "260px";

    return (
      <Tooltip {...rest} style={{ width: maxWidth, padding: 0 }}>
        {row !== undefined &&
          seat !== undefined &&
          section !== undefined && (
            <SeatData size={size} hasChildren={!!children}>
              <SeatDataColumn>
                <ColumnHeading size={size}>Sec</ColumnHeading>
                <ColumnText variant={rest.variant} size={size}>
                  {section}
                </ColumnText>
              </SeatDataColumn>
              <SeatDataColumn>
                <ColumnHeading size={size}>Row</ColumnHeading>
                <ColumnText variant={rest.variant} size={size}>
                  {row}
                </ColumnText>
              </SeatDataColumn>
              <SeatDataColumn>
                <ColumnHeading size={size}>Seat</ColumnHeading>
                <ColumnText variant={rest.variant} size={size}>
                  {seat}
                </ColumnText>
              </SeatDataColumn>
            </SeatData>
          )}
        {children && (
          <AdditionalData size={size} variant={rest.variant}>
            {children}
          </AdditionalData>
        )}
      </Tooltip>
    );
  }
}

SeatTooltip.propTypes = {
  section: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  row: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  seat: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  size: PropTypes.oneOf(Object.values(seatTooltipSizes)),
  children: PropTypes.node
};

SeatTooltip.defaultProps = {
  size: seatTooltipSizes.large,
  children: null
};

SeatTooltip.displayName = "SeatTooltip";

export default SeatTooltip;
