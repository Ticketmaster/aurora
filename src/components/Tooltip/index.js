import React, { Component } from "react";
import PropTypes from "prop-types";
import { CSSTransition } from "react-transition-group";
import Portal from "../PopOver/PopOverPortal";
import PopOver from "../PopOver/index";
import { StyledTooltip } from "./Tooltip.styles";
import SPACE_FROM_MOUSE from "./constants";
import {
  DIRECTIONS,
  TOP,
  LEFT,
  RIGHT,
  BOTTOM,
  VARIANTS,
  LIGHT,
  AUTO,
  ARROW_WIDTH
} from "../constants";
import { constants } from "../../theme";

class Tooltip extends Component {
  constructor(props) {
    super(props);

    this.myRef = React.createRef();
    this.dimensions = {
      width: 0,
      height: 0,
      windowScroll: 0,
      windowWidth: 0,
      windowHeight: 0
    };

    this.pos = {
      x: 0,
      y: 0
    };

    this.actualDirection = props.direction;

    this.state = {
      actualDirection: props.direction,
      arrowAdjustment: 0
    };
  }

  /*
   * Static function that needs to be called from the parent -> Tooltip.getDimensionsFromEvent
   * The parent should pass the click event which will trigger showing the Tooltip.
   * By default the Tooltip is shown withing the view port. If we need to show it inside
   * certain element we should pass the html element as second parameter.
   * The function will return an object that should be provided to the Tooltip as props.
   */
  static getDimensionsFromEvent(e, parent) {
    const {
      y: elTop,
      height: elHeight,
      x: elLeft,
      width: elWidth
    } = e.target.getBoundingClientRect();
    const {
      y: offsetTop = 0,
      height: clientHeight = 100000,
      x: offsetLeft = 0,
      width: clientWidth = 100000
    } =
      parent && parent.getBoundingClientRect
        ? parent.getBoundingClientRect()
        : {};

    return {
      elTop,
      elLeft,
      elWidth,
      elBottom: elTop + elHeight,
      elRight: elLeft + elWidth,
      offsetTop,
      clientHeight,
      offsetLeft,
      clientWidth,
      elHorizontalCenter: elLeft + elWidth / 2,
      elVerticalCenter: elTop + elHeight / 2
    };
  }

  getPositionAndUpdateDirection = ({
    position,
    dimensions,
    spaceFromMouse,
    reduce
  }) => {
    const result = PopOver.calculatePosition({
      position,
      dimensions,
      spaceFromMouse,
      reduce
    });
    const { actualDirection, arrowAdjustment } = this.state;

    const adjustment = this.adjustArrow({
      coords: { x: result.x, width: dimensions.width },
      position
    });

    if (
      result.y < position.elTop + dimensions.windowScroll &&
      (actualDirection !== TOP || arrowAdjustment !== adjustment)
    ) {
      this.actualDirection = TOP;
      this.setState({ actualDirection: TOP, arrowAdjustment: adjustment });
    } else if (
      result.y > position.elTop + dimensions.windowScroll &&
      (actualDirection !== BOTTOM || arrowAdjustment !== adjustment)
    ) {
      this.actualDirection = BOTTOM;
      this.setState({ actualDirection: BOTTOM, arrowAdjustment: adjustment });
    }

    return result;
  };

  getTranslateByDirection = direction => {
    switch (direction) {
      case TOP:
        return "translate(0, 10px)";
      case BOTTOM:
        return "translate(0, -10px)";
      case LEFT:
        return "translate(10px, 0)";
      case RIGHT:
      default:
        return "translate(-10px, 0)";
    }
  };

  adjustArrow = ({ coords, position }) => {
    const reqCenter = position.elHorizontalCenter;
    const currentCenter = coords.x + coords.width / 2;
    const mostLeft = coords.x + ARROW_WIDTH / 2;
    const mostRight = coords.x + coords.width + ARROW_WIDTH / 2;
    const center = Math.min(Math.max(mostLeft, reqCenter), mostRight);

    return center - currentCenter;
  };

  /*
   * Function that determines Tooltip position.
   * @position(object) - top and bottom position of the elemt that triggers showing Tooltip;
   * Mouse horizontal position on the sreen - so we can center the Tooltip;
   * Additional container position and size
   * @size(object) - Tooltip width and height;
   */
  calculatePosition = ({
    direction,
    position,
    dimensions,
    spaceFromMouse,
    reduce
  }) => {
    const { width, height, windowScroll } = dimensions;
    const { elBottom, elTop, elLeft, elRight, elHorizontalCenter } = position;

    const bottomPosition = elBottom + SPACE_FROM_MOUSE;
    const topPosition = elTop - SPACE_FROM_MOUSE - height + windowScroll;

    switch (direction) {
      case TOP:
        return {
          x: elHorizontalCenter - width / 2,
          y: topPosition
        };
      case BOTTOM:
        return {
          x: elHorizontalCenter - width / 2,
          y: bottomPosition + windowScroll
        };
      case LEFT:
        return {
          x: elLeft - width - SPACE_FROM_MOUSE,
          y: elTop + windowScroll
        };
      case RIGHT:
        return {
          x: elRight + SPACE_FROM_MOUSE,
          y: elTop + windowScroll
        };
      default:
        return this.getPositionAndUpdateDirection({
          direction,
          position,
          dimensions,
          spaceFromMouse,
          reduce
        });
    }
  };

  updateSize = () => {
    const { isVisible } = this.props;
    const {
      windowScroll,
      windowWidth,
      windowHeight,
      width,
      height
    } = this.dimensions;
    const dimensions = {};
    if (global.window && isVisible) {
      const {
        clientWidth,
        clientHeight
      } = global.window.document.documentElement;

      const scrollTop = Math.max(
        global.window.pageYOffset,
        global.document.documentElement.scrollTop,
        global.document.body.scrollTop
      );

      if (scrollTop !== windowScroll) {
        dimensions.windowScroll = scrollTop;
      }

      if (clientWidth !== windowWidth) {
        dimensions.windowWidth = clientWidth;
      }

      if (clientHeight !== windowHeight) {
        dimensions.windowHeight = clientHeight;
      }
    }

    if (this.myRef.current) {
      const { clientWidth, clientHeight } = this.myRef.current;

      if (width !== clientWidth && clientWidth) {
        dimensions.width = clientWidth;
      }

      if (height !== clientHeight && clientHeight) {
        dimensions.height = clientHeight;
      }
    }

    if (Object.keys(dimensions).length) {
      this.dimensions = {
        ...this.dimensions,
        ...dimensions
      };
      return true;
    }

    return false;
  };

  tooltipEnter = () => {
    const {
      isVisible,
      position,
      direction,
      spaceFromMouse,
      reduceTop,
      reduceBottom
    } = this.props;

    if (isVisible) {
      const reduce = { top: reduceTop, bottom: reduceBottom };
      this.updateSize();
      this.pos = this.calculatePosition({
        direction,
        position,
        dimensions: this.dimensions,
        spaceFromMouse,
        reduce
      });
    }

    this.myRef.current.style.top = `${this.pos.y}px`;
    this.myRef.current.style.left = `${this.pos.x}px`;
    this.myRef.current.style.transition = `opacity 0.3s ${
      constants.easing.easeOutQuad
    }`;
    this.myRef.current.style.transform = this.getTranslateByDirection(
      this.actualDirection
    );
  };

  tooltipEntering = () => {
    this.myRef.current.style.transition = `opacity 0.3s ${
      constants.easing.easeOutQuad
    },
      transform 0.3s ${constants.easing.easeOutQuad}`;
    this.myRef.current.style.transform = "translate(0)";
  };

  tooltipExit = () => {
    this.myRef.current.style.transition = `opacity 0.1s ${
      constants.easing.easeInQuad
    }`;
  };

  render() {
    const { children, isVisible, variant, ...rest } = this.props;
    const { actualDirection, arrowAdjustment } = this.state;

    return (
      <Portal>
        <CSSTransition
          in={isVisible}
          key="tooltip-animation"
          timeout={300}
          classNames="open"
          onEnter={this.tooltipEnter}
          onEntering={this.tooltipEntering}
          onExit={this.tooltipExit}
          appear={isVisible}
          variant={variant}
        >
          <StyledTooltip
            ref={this.myRef}
            isVisible={isVisible}
            {...rest}
            direction={actualDirection}
            arrowAdjustment={`${arrowAdjustment}px`}
          >
            {children}
          </StyledTooltip>
        </CSSTransition>
      </Portal>
    );
  }
}

Tooltip.propTypes = {
  children: PropTypes.node.isRequired,
  isVisible: PropTypes.bool,
  direction: PropTypes.oneOf(DIRECTIONS),
  position: PropTypes.shape({
    elHorizontalCenter: PropTypes.number,
    elVerticalCenter: PropTypes.number,
    elTop: PropTypes.number,
    elBottom: PropTypes.number,
    elLeft: PropTypes.number
  }),
  variant: PropTypes.oneOf(VARIANTS),
  spaceFromMouse: PropTypes.number,
  reduceTop: PropTypes.number,
  reduceBottom: PropTypes.number
};

Tooltip.defaultProps = {
  isVisible: false,
  direction: AUTO,
  variant: LIGHT,
  position: {
    mouseX: 0,
    elTop: 0,
    elBottom: 0,
    offsetTop: 0,
    clientHeight: 0,
    offsetLeft: 0,
    clientWidth: 0
  },
  spaceFromMouse: SPACE_FROM_MOUSE,
  reduceTop: 0,
  reduceBottom: 0
};

Tooltip.displayName = "Tooltip";

export default Tooltip;
