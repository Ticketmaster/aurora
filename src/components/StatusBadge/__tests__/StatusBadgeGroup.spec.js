import React from "react";
import renderer from "react-test-renderer";
import "jest-styled-components";
import { StatusBadgeGroup } from "../";
import Tooltip from "../../Tooltip";

const badges = [
  { label: "error", color: "#d3d3d3" },
  { label: "caution", color: "#f0f0f0" }
];

describe("StatusBadgeGroup", () => {
  it("should match snapshot", () => {
    const tree = renderer
      .create(
        <StatusBadgeGroup
          variant="dark"
          visibleBadges={badges}
          hiddenBadges={badges}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("mouseLeave should set state to isOpen: false", () => {
    const setStateMock = jest.fn();
    const instance = renderer
      .create(
        <StatusBadgeGroup
          variant="dark"
          visibleBadges={badges}
          hiddenBadges={badges}
        />
      )
      .getInstance();

    instance.setState = setStateMock;
    instance.mouseLeave();

    expect(setStateMock).toHaveBeenCalledTimes(1);
    expect(setStateMock).toHaveBeenCalledWith({
      isOpen: false
    });
  });

  it("elementHovered should set state to isOpen: true and position data", () => {
    const setStateMock = jest.fn();
    Tooltip.getDimensionsFromEvent = jest.fn(() => ({
      test: "test data"
    }));
    const instance = renderer
      .create(
        <StatusBadgeGroup
          variant="dark"
          visibleBadges={badges}
          hiddenBadges={badges}
        />
      )
      .getInstance();

    instance.setState = setStateMock;
    instance.elementHovered();

    expect(setStateMock).toHaveBeenCalledTimes(1);
    expect(Tooltip.getDimensionsFromEvent).toHaveBeenCalledTimes(1);
    expect(setStateMock).toHaveBeenCalledWith({
      isOpen: true,
      test: "test data"
    });
  });
});
