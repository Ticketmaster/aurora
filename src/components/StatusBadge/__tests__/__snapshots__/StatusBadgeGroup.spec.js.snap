// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusBadgeGroup should match snapshot 1`] = `
.c1 {
  padding: 4px 8px;
  font-weight: 600;
  text-align: center;
  color: rgba(255,255,255,1);
  font-size: 10px;
  line-height: 1.2;
  border-radius: 10px;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  text-transform: uppercase;
  background-color: #d3d3d3;
}

.c2 {
  padding: 4px 8px;
  font-weight: 600;
  text-align: center;
  color: rgba(255,255,255,1);
  font-size: 10px;
  line-height: 1.2;
  border-radius: 10px;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  text-transform: uppercase;
  background-color: #f0f0f0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 .badge {
  margin-left: 8px;
}

.c0 .badge:first-child,
.c0 .badge.hidden {
  margin-left: 0;
}

.c0 .badge.hidden {
  margin-top: 8px;
}

.c0 .badge.hidden:first-child {
  margin-top: 0;
}

.c3 {
  padding: 4px 8px;
  font-weight: 600;
  text-align: center;
  color: rgba(255,255,255,1);
  font-size: 10px;
  line-height: 1.2;
  border-radius: 10px;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  text-transform: uppercase;
  background: transparent;
  color: rgba(38,38,38,0.65);
}

<div
  className="c0"
>
  <div
    className="c1 badge"
    color="#d3d3d3"
  >
    error
  </div>
  <div
    className="c2 badge"
    color="#f0f0f0"
  >
    caution
  </div>
  <div
    className="c3"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onTouchStart={[Function]}
  >
    +2 more
  </div>
</div>
`;
